spy/vm/astframe.py

```diff
@@ -803,6 +803,39 @@ def eval_expr_CmpOp(self, op: ast.CmpOp) -> W_MetaArg:
         w_opimpl = self.vm.call_OP(op.loc, w_OP, [wam_l, wam_r])
         return self.eval_opimpl(op, w_opimpl, [wam_l, wam_r])
 
+    def eval_expr_CmpChain(self, chain: ast.CmpChain) -> W_MetaArg:
+        assert chain.comparisons, "CmpChain must contain at least one comparison"
+        first_cmp = chain.comparisons[0]
+        prev_expr = first_cmp.left
+        wam_prev = self.eval_expr(prev_expr)
+        result: Optional[W_MetaArg] = None
+        saw_red = False
+
+        for cmp in chain.comparisons:
+            assert cmp.left is prev_expr, "chained comparisons must share operands"
+            wam_right = self.eval_expr(cmp.right)
+            w_OP = OP_from_token(cmp.op)
+            w_opimpl = self.vm.call_OP(cmp.loc, w_OP, [wam_prev, wam_right])
+            result = self.eval_opimpl(cmp, w_opimpl, [wam_prev, wam_right])
+            if result.color == "red":
+                saw_red = True
+            if result._w_val is not None:
```

Comment:
this check is also suspicious, I don't think that checking for `_w_val` is the right thing to do.
I **think** that the right logic is:
  - if we are redshifting, we want to evaluate all the operands without bothering to check their result (because we are doing abstrat interpretation and we want to ensure that we are evaluating all possible paths)
  - if we are not redshifting, then we *know for sure* that `result._w_val` is not None, and we can just access `result.w_val`.